src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -0,0 +1,751 @@
+                static_cast<std::uint32_t>(available.size() - 1))];
+    }
+
+    // Large, sparse index space to sample
+    // Keep choosing indexes at random until an available one is found
+    // chances of running more than 30 times is less than 1 in a billion
+    for (int i = 0; i < 40; ++i)
+    {
+        auto const r = rand_int(genesisShardIndex, maxShardIndex);
+        if (complete_.find(r) == complete_.end() &&
+            (!incomplete_ || incomplete_->index() != r))
+                return r;
+    }
+    assert(0);
+    return boost::none;
+}
+
+void
+DatabaseShardImp::updateStats(std::lock_guard<std::mutex>&)
+{
+    // Calculate shard file sizes and update status string
+    std::uint32_t filesPerShard {0};
+    if (!complete_.empty())
+    {
+        status_.clear();
+        filesPerShard = complete_.begin()->second->fdlimit();
+        std::uint64_t avgShardSz {0};
+        for (auto it = complete_.begin(); it != complete_.end(); ++it)
+        {
+            if (it == complete_.begin())
+                status_ = std::to_string(it->first);
+            else
+            {
+                if (it->first - std::prev(it)->first > 1)
+                {
+                    if (status_.back() == '-')
+                        status_ += std::to_string(std::prev(it)->first);
+                    status_ += "," + std::to_string(it->first);
+                }
+                else
+                {
+                    if (status_.back() != '-')
+                        status_ += "-";
+                    if (std::next(it) == complete_.end())
+                        status_ += std::to_string(it->first);
+                }
+            }
+            avgShardSz += it->second->fileSize();
+        }
+        if (backed_)
+            avgShardSz_ = avgShardSz / complete_.size();
+    }
+    else if(incomplete_)
+        filesPerShard = incomplete_->fdlimit();
+    if (!backed_)
+        return;
+
+    fdLimit_ = 1 + (filesPerShard *
+        (complete_.size() + (incomplete_ ? 1 : 0)));
+
+    if (usedDiskSpace_ >= maxDiskSpace_)
+    {
+        JLOG(j_.warn()) <<
+            "Maximum size reached";
+        canAdd_ = false;
+    }
+    else
+    {
+        auto const sz = maxDiskSpace_ - usedDiskSpace_;
+        if (sz > space(dir_).free)
+        {
+            JLOG(j_.warn()) <<
+                "Max Shard Store size exceeds "
+                "remaining free disk space";
+        }
+        fdLimit_ += (filesPerShard * (sz / avgShardSz_));
+    }
+}
+
+boost::optional<std::pair<std::shared_ptr<PCache>,
+    std::shared_ptr<NCache>>>
+DatabaseShardImp::selectCache(std::uint32_t seq,
+    std::lock_guard<std::mutex>&)
+{
+    boost::optional<std::pair<std::shared_ptr<PCache>,
+        std::shared_ptr<NCache>>> cache;
+    auto const shardIndex {seqToShardIndex(seq)};
+    auto it = complete_.find(shardIndex);
+    if (it != complete_.end())
+    {
+        cache->first = it->second->pCache();
```

Comment:
You need to initialize the optional before assigning to the pair members, or just do a `cache = std::make_pair`.  This crashed for me when running with `assert`.

Reply:
Good catch, fixed.